client/minimal-relay-chain-node/src/collator_overseer.rs

```diff
@@ -0,0 +1,246 @@
+			.approval_distribution(DummySubsystem)
+			.approval_voting(DummySubsystem)
+			.gossip_support(DummySubsystem)
+			.dispute_coordinator(DummySubsystem)
+			.dispute_distribution(DummySubsystem)
+			.chain_selection(DummySubsystem)
+			.leaves(Vec::from_iter(
+				leaves
+					.into_iter()
+					.map(|BlockInfo { hash, parent_hash: _, number }| (hash, number)),
+			))
+			.activation_external_listeners(Default::default())
+			.span_per_active_leaf(Default::default())
+			.active_leaves(Default::default())
+			.supports_parachains(runtime_client)
+			.known_leaves(LruCache::new(KNOWN_LEAVES_CACHE_SIZE))
+			.metrics(metrics)
+			.spawner(spawner);
+
+		builder.build_with_connector(connector).map_err(|e| e.into())
+	}
+}
+
+pub(crate) fn spawn_overseer(
+	overseer_args: CollatorOverseerGenArgs,
+	task_manager: &TaskManager,
+	relay_chain_rpc_client: Arc<BlockChainRpcClient>,
+) -> Result<polkadot_overseer::Handle, polkadot_service::Error> {
+	let overseer_gen = CollatorOverseerGen;
+	let (overseer, overseer_handle) = overseer_gen
+		.generate(OverseerConnector::default(), overseer_args)
+		.map_err(|e| {
+			tracing::error!("Failed to initialize overseer: {}", e);
+			e
+		})?;
+
+	let overseer_handle = Handle::new(overseer_handle.clone());
+	{
+		let handle = overseer_handle.clone();
+		task_manager.spawn_essential_handle().spawn_blocking(
+			"overseer",
+			None,
+			Box::pin(async move {
+				use futures::{pin_mut, FutureExt};
+
+				let forward = forward_collator_events(relay_chain_rpc_client, handle).fuse();
+
+				// let forward = forward.fuse();
+				let overseer_fut = overseer.run().fuse();
+
+				pin_mut!(overseer_fut);
+				pin_mut!(forward);
+
+				select! {
+					_ = forward => (),
+					_ = overseer_fut => (),
+					complete => (),
+				}
+			}),
+		);
+	}
+	Ok(overseer_handle)
+}
+
+pub struct NewCollator {
+	pub task_manager: TaskManager,
+	pub overseer_handle: Handle,
+	pub network: Arc<sc_network::NetworkService<Block, <Block as BlockT>::Hash>>,
+}
+
+/// Glues together the [`Overseer`] and `BlockchainEvents` by forwarding
+/// import and finality notifications into the [`OverseerHandle`].
+pub(crate) async fn forward_collator_events(client: Arc<BlockChainRpcClient>, mut handle: Handle) {
+	let mut finality = client.finality_notification_stream_async().await.fuse();
+	let mut imports = client.import_notification_stream_async().await.fuse();
+
+	loop {
+		select! {
+			f = finality.next() => {
+				match f {
+					Some(header) => {
+		let block_info = BlockInfo { hash: header.hash(), parent_hash: header.parent_hash, number: header.number };
+						handle.block_finalized(block_info).await;
+					}
+					None => break,
+				}
+			},
+			i = imports.next() => {
+				match i {
+					Some(header) => {
+		let block_info = BlockInfo { hash: header.hash(), parent_hash: header.parent_hash, number: header.number };
+						handle.block_imported(block_info).await;
+					}
+					None => break,
+				}
+			},
+			complete => break,
```

Comment:
```suggestion
```
Maybe this is not strictly required as the loop terminates as soon as one of the two futures yields `None`